### PR TITLE
WIP: Hook support

### PR DESCRIPTION
**Pull request by davejlong**
Adds support for webhooks in Kitto. More documentation to be provided as the PR is completed.

### CODE REVIEW

File: lib/kitto/hook.ex [hunk 1]
```diff
@@ -0,0 +1,11 @@
+defmodule Kitto.Hook do
+  @moduledoc """
+  Contains functions to run hooks based on their specified options.
+  """
+
+  alias Kitto.Runner
+
+  def register(server, name, options, definition, hook) do
+    Runner.register server, {:hook, %{name: to_string(name), hook: hook, options: options, definition: definition}}
```
Comment (zorbash) [inline lib/kitto/hook.ex]: Line is too long (max is 90, was 115).

File: lib/kitto/hook.ex [hunk 2]
```diff
@@ -0,0 +1,11 @@
+defmodule Kitto.Hook do
+  @moduledoc """
+  Contains functions to run hooks based on their specified options.
+  """
+
+  alias Kitto.Runner
+
+  def register(server, name, options, definition, hook) do
```
Comment (zorbash) [inline lib/kitto/hook.ex]: Functions should have a @spec type specification.

File: lib/kitto/runner.ex [hunk 3]
```diff
@@ -21,16 +21,16 @@ defmodule Kitto.Runner do
...
+    {:ok, %{opts: opts, jobs: [], hooks: %{}, supervisor: nil}}
   end
 
   @doc """
   Updates the list of jobs to be run with the provided one
   """
-  def register(server, job) do
-    GenServer.call(server, {:register, job})
-  end
+  def register(server, {:job, job}), do: GenServer.call(server, {:register_job, job})
```
Comment (Zorbash) [inline lib/kitto/runner.ex]: Functions should have a @spec type specification.

File: lib/kitto/runner.ex [hunk 4]
```diff
@@ -21,16 +21,16 @@ defmodule Kitto.Runner do
...
+  def register(server, {:hook, hook}), do: GenServer.call(server, {:register_hook, hook})
```
Comment (Zorbash) [inline lib/kitto/runner.ex]: Functions should have a @spec type specification.

File: lib/kitto/runner.ex [hunk 5]
```diff
@@ -54,20 +54,45 @@ defmodule Kitto.Runner do
   end
 
   @doc """
+  Returns all the registered hooks
+  """
+  def hooks(server) do
```
Comment (Zorbash) [inline lib/kitto/runner.ex]: Functions should have a @spec type specification.

File: lib/kitto/runner.ex [hunk 6]
```diff
@@ -54,20 +54,45 @@ defmodule Kitto.Runner do
...
+    GenServer.call(server, {:hooks})
+  end
+
+  @doc """
+  Gets a specific hook from the registry
+  """
+  def hook(server, id) do
```
Comment (Zorbash) [inline lib/kitto/runner.ex]: Functions should have a @spec type specification.

File: lib/kitto/runner.ex [hunk 7]
```diff
@@ -54,20 +54,45 @@ defmodule Kitto.Runner do
...
+  Returns the directory where the hook scripts are located
+  """
+  def hooks_dir, do: dir Application.get_env(:kitto, :hooks_dir, "hooks")
 
   ### Callbacks
 
-  def handle_call({:jobs}, _from, state) do
-    {:reply, state.jobs, state}
-  end
+  def handle_call({:jobs}, _from, state), do: {:reply, state.jobs, state}
```
Comment (Zorbash) [inline lib/kitto/runner.ex]: Functions should have a @spec type specification.

File: lib/kitto/runner.ex [hunk 8]
```diff
@@ -54,20 +54,45 @@ defmodule Kitto.Runner do
...
+
+  def handle_call({:hooks}, _from, state), do: {:reply, state.hooks, state}
```
Comment (Zorbash) [inline lib/kitto/runner.ex]: Functions should have a @spec type specification.

File: lib/kitto/runner.ex [hunk 9]
```diff
@@ -54,20 +54,45 @@ defmodule Kitto.Runner do
...
 
-  def handle_call({:register, job}, _from, state) do
+  def handle_call({:hook, id}, _from, state), do: {:reply, state.hooks[to_string(id)], state}
```
Comment (Zorbash) [inline lib/kitto/runner.ex]: Functions should have a @spec type specification.

File: lib/kitto/runner.ex [hunk 10]
```diff
@@ -54,20 +54,45 @@ defmodule Kitto.Runner do
...
+
+  def handle_call({:register_job, job}, _from, state) do
```
Comment (Zorbash) [inline lib/kitto/runner.ex]: Functions should have a @spec type specification.

File: lib/kitto/runner.ex [hunk 11]
```diff
@@ -54,20 +54,45 @@ defmodule Kitto.Runner do
...
     {:reply, job, %{state | jobs: state.jobs ++ [job]}}
   end
 
+  def handle_call({:register_hook, hook}, _from, state) do
```
Comment (Zorbash) [inline lib/kitto/runner.ex]: Functions should have a @spec type specification.

File: lib/kitto/runner.ex [hunk 12]
```diff
@@ -78,6 +103,12 @@ defmodule Kitto.Runner do
    {:noreply, %{state | supervisor: supervisor}}
   end
 
+  def handle_cast({:hooks_loaded}, state) do
```
Comment (Zorbash) [inline lib/kitto/runner.ex]: Functions should have a @spec type specification.

File: lib/kitto/runner.ex [hunk 13]
```diff
@@ -136,5 +164,24 @@ defmodule Kitto.Runner do
...
   end
 
-  defp job_files, do: Path.wildcard(Path.join(jobs_dir(), "/**/*.{ex,exs}"))
+  defp load_hooks(pid) do
+    hook_files() |> Enum.each(&(load_hook(pid, &1)))
+
+    GenServer.cast pid, {:hooks_loaded}
+  end
+
+  def load_file(pid, file, type) do
```
Comment (Zorbash) [inline lib/kitto/runner.ex]: Functions should have a @spec type specification.

File: lib/kitto/job.ex [hunk 14]
```diff
@@ -24,7 +24,7 @@ defmodule Kitto.Job do
     opts = [interval: options[:every] |> mseconds,
             first_at: options[:first_at] |> mseconds]
 
-    Runner.register server, %{name: name, job: job, options: opts, definition: definition}
+    Runner.register server, {:job, %{name: name, job: job, options: opts, definition: definition}}
```
Comment (zorbash) [inline lib/kitto/job.ex:27]: Line is too long (max is 90, was 98).

**Comment by Zorbash** [comment]
Ebert has finished reviewing this Pull Request and has found:

* 4 possible new issues (including those that may have been commented here).

But beware that this branch is **[2 commits behind the `kittoframework:master` branch](https://github.com/kittoframework/kitto/compare/kittoframework:hook-support...kittoframework:master)**, and a review of an up to date branch would produce more accurate results.

You can see more details about this review at https://ebertapp.io/github/kittoframework/kitto/pulls/95.

**Comment by davejlong** [comment]
I've worked on creating a second attempt at adding support for webhooks to Kitto. This is still WIP, but I am not sure how to accomplish what needs to be done. In the new hook router that I've created, `lib/kitto/hook/router.ex`, I need to use `Kitto.Runner` to get access to hooks. From my test code, I am going to spin up a new instance of the `Runner` so that I can control the hooks that are loaded into it, but I can't use the same namespace as the `Runner` which is always running in Kitto. Furthermore, I can't pass options into the router because of a limitation with `Plug.Router`.

I'm looking for any advice on how best to change out the `Kitto.Runner` registry when testing the router.

**Comment by davejlong** [comment]
I'm thinking of rewriting the implementation a bit as follows:

Split hooks into it's own registry. Hooks and jobs have very different functionality and trying to cover them both in `Kitto.Runner` is turning the module into giant monolithic mess.

The second change is to merge `Kitto.Hook.DSL` and `Kitto.Job.DSL` into `Kitto.DSL`. Similar to Phoenix's `MyProject.Web` module, someone writing a hook would `use Kitto.DSL, :hook` and someone writing a job would `use Kitto.DSL, :job`. I think that I can achieve backwards compatibility to support `Kitto.Job.DSL` which would just `use Kitto.DSL, :job`, but I'm not sure if we need that given that we are not 1.0 yet.

File: lib/kitto/runner.ex [hunk 15]
```diff
@@ -21,16 +21,16 @@ defmodule Kitto.Runner do
...
     server = self()
     spawn fn -> load_jobs(server) end
+    spawn fn -> load_hooks(server) end
 
-    {:ok, %{opts: opts, jobs: [], supervisor: nil}}
+    {:ok, %{opts: opts, jobs: [], hooks: %{}, supervisor: nil}}
   end
 
   @doc """
   Updates the list of jobs to be run with the provided one
```
Comment (zorbash) [inline lib/kitto/runner.ex:28]: We'll also have to update the doc here.

**Comment by zorbash** [comment]
@davejlong Concerning https://github.com/kittoframework/kitto/pull/95#issuecomment-288264387 I'm positive for your rewrite plan. Let's also have `use Kitto.DSL` keep compatiblity with existing jobs if possible.

**Pull request closed without merging**